### PR TITLE
Improve popup window/menu handling.

### LINKED ISSUE #53187: Editor window title is flickering when using popups/menu with single window mode off (DisplayServer regression)
### Godot version

Master branch (84b11fbaa674b6e2341de534e93a981378a307ad)

### System information

Windows10 , Linux Ubuntu 21.04

### Issue description

The main editor window loose focus when using menus or popup in multi window mode.
That cause the title flickering between faded state and normal state which is particulary visible using the main editor menu.

focus

menu_focus_linux
May be related to #44997

### Steps to reproduce

Use main editor menu when single window mode is disabled in editor properties as in the gif below.

### Minimal reproduction project

Not needed.

### PR DESCRIPTION
**Pull request by bruvzg**
Adds a dedicated window `FLAG_POPUP` and a platform specific routines to control popup menus.

<details>
<summary>Previews:</summary>

macOS:

https://user-images.githubusercontent.com/7645683/155496415-5fb0ed4e-0c47-4edd-8368-a8da07683c54.mov

Windows:

https://user-images.githubusercontent.com/7645683/155511747-7fb51328-60fc-4865-8730-6937cf9ac524.mov

</details>

## TODO:
- [x] macOS implementation.
- [x] Windows implementation.
- [x] Linux/X11 implementation.
- [x] Documentation update.
- [x] Test `Popup`s other than menu and `OptionButton` dropdowns are functioning correctly.

Fixes #53187, supersede #58461

### CODE REVIEW

**Comment by akien-mga** [comment]
Tested on Linux X11 with Plasma/KWin, seems to work fine.
Would be good to have it tested on Wayland and maybe on a tiled WM (@groud maybe?).

**Comment by Geometror** [comment]
Tested on Windows 10 (since bruvzg screenshot was taken in W11) and it seems to work fine. It also fixes an important regression where all context menus wouldn't show up anymore after a tooltip is displayed (not mentioned in #58515).

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

File: platform/linuxbsd/display_server_x11.cpp [hunk 1]
```diff
@@ -2073,6 +2076,18 @@ void DisplayServerX11::window_set_flag(WindowFlags p_flag, bool p_enabled, Windo
...
+		case WINDOW_FLAG_NO_FOCUS: {
+			wd.no_focus = p_enabled;
+		} break;
+		case WINDOW_FLAG_POPUP: {
+			XWindowAttributes xwa;
+			XSync(x11_display, False);
+			XGetWindowAttributes(x11_display, wd.x11_window, &xwa);
+
+			ERR_FAIL_COND_MSG(p_window == MAIN_WINDOW_ID, "Main window can't be popup.");
+			ERR_FAIL_COND_MSG((xwa.map_state == IsViewable) && (wd.is_popup != p_enabled), "Pupup flag can't changed while window is opened.");
```
Comment (akien-mga) [inline platform/linuxbsd/display_server_x11.cpp:2088]: There's a typo here in "P**u**pup" (and other 2 implementations).

File: platform/linuxbsd/display_server_x11.cpp [hunk 2]
```diff
@@ -1458,8 +1461,8 @@ void DisplayServerX11::window_set_transient(WindowID p_window, WindowID p_parent
 		// Set focus to parent sub window to avoid losing all focus when closing a nested sub-menu.
 		// RevertToPointerRoot is used to make sure we don't lose all focus in case
 		// a subwindow and its parent are both destroyed.
-		if (wd_window.menu_type && !wd_window.no_focus && wd_window.focused) {
-			if (!wd_parent.no_focus) {
+		if (!wd_window.no_focus && !wd_window.is_popup && wd_window.focused) {
+			if (!wd_parent.no_focus && !wd_window.is_popup) {
```
Comment (Rindbee) [inline platform/linuxbsd/display_server_x11.cpp:1465]: `!wd_window.is_popup` appears twice. The second is likely to be `!wd_parent.is_popup`.

**Pull request merged**